File: AgoraChat/Code/Custom/ViewController/ACDSearchTableViewController.m
//
//  AgoraSearchTableViewController.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/10/21.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDSearchTableViewController.h"
#import "AgoraRealtimeSearchUtils.h"

#define kSearchBarHeight 32.0

@interface ACDSearchTableViewController ()<UISearchBarDelegate>
@property (nonatomic, strong) NSMutableArray *searchResults;
@property (nonatomic, assign) BOOL isSearchState;

@end

@implementation ACDSearchTableViewController
- (void)viewDidLoad {
    [super viewDidLoad];
}

- (void)prepare {
    [super prepare];
    [self.view addSubview:self.searchBar];
}

- (void)placeSubViews {
    [self.searchBar mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.view);
        make.left.and.right.equalTo(self.view);
        make.height.mas_equalTo(kSearchBarHeight);
    }];
    
    [self.table mas_makeConstraints:^(MASConstraintMaker* make) {
        make.top.equalTo(self.searchBar.mas_bottom);
        make.left.and.right.equalTo(self.view);
        make.bottom.equalTo(self.view).offset(-5.0);

    }];

}


#pragma mark - UISearchBarDelegate
- (BOOL)searchBarShouldBeginEditing:(UISearchBar *)searchBar {
    [searchBar setShowsCancelButton:YES animated:YES];
    return YES;
}

- (void)searchBarTextDidBeginEditing:(UISearchBar *)searchBar {

    _isSearchState = YES;
    if ([searchBar.text isEqualToString:@""]) {
        self.table.userInteractionEnabled = NO;
        [self.searchResults removeAllObjects];
        [self.table reloadData];
    }else {
        self.table.userInteractionEnabled = YES;
    }
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar {
    [searchBar setShowsCancelButton:NO animated:YES];
    self.table.userInteractionEnabled = YES;
    [self.searchBar resignFirstResponder];
}

- (BOOL)searchBar:(UISearchBar *)searchBar shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text {
    return YES;
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    self.table.userInteractionEnabled = YES;
    
    [_searchResults removeAllObjects];
    if (searchBar.text.length == 0) {
        [self.table reloadData];
        return;
    }
    
    __weak typeof(self) weakSelf = self;
    [[AgoraRealtimeSearchUtils defaultUtil] realtimeSearchWithSource:_searchSource searchString:searchText resultBlock:^(NSArray *results) {
        if (results) {
            dispatch_async(dispatch_get_main_queue(), ^{
                _searchResults = [NSMutableArray arrayWithArray:results];
                if (weakSelf.searchResultBlock) {
                    weakSelf.searchResultBlock();
                }
                
                [weakSelf.table reloadData];
            });
        }
    }];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
    [searchBar resignFirstResponder];
}

- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar {
    searchBar.text = @"";
    [searchBar setShowsCancelButton:NO animated:NO];
    [searchBar resignFirstResponder];
    [[AgoraRealtimeSearchUtils defaultUtil] realtimeSearchDidFinish];
    _isSearchState = NO;
    if (self.searchCancelBlock) {
        self.searchCancelBlock();
    }
    [self.searchResults removeAllObjects];
    self.table.scrollEnabled = !_isSearchState;
    [self.table reloadData];
}

#pragma mark public method
- (void)cancelSearchState {
    self.searchBar.text = @"";
    [self.searchBar setShowsCancelButton:NO animated:NO];
    [self.searchBar resignFirstResponder];
    [[AgoraRealtimeSearchUtils defaultUtil] realtimeSearchDidFinish];
    _isSearchState = NO;
    if (self.searchCancelBlock) {
        self.searchCancelBlock();
    }
    [self.searchResults removeAllObjects];
    self.table.scrollEnabled = !_isSearchState;
    [self.table reloadData];
}

- (void)sortContacts:(NSArray *)contacts {
    if (contacts.count == 0) {
        self.dataArray = [@[] mutableCopy];
        self.sectionTitles = [@[] mutableCopy];
        self.searchSource = [@[] mutableCopy];
        return;
    }
    
    NSMutableArray *sectionTitles = nil;
    NSMutableArray *searchSource = nil;
    NSArray *sortArray = [NSArray sortContacts:contacts
                                 sectionTitles:&sectionTitles
                                  searchSource:&searchSource];
    [self.dataArray removeAllObjects];
    [self.dataArray addObjectsFromArray:sortArray];
    self.sectionTitles = [NSMutableArray arrayWithArray:sectionTitles];
    self.searchSource = [NSMutableArray arrayWithArray:searchSource];
}

#pragma mark getter and setter
- (UISearchBar*)searchBar
{
    if (_searchBar == nil) {
        _searchBar = [[UISearchBar alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, kSearchBarHeight)];
        _searchBar.placeholder = @"Search";
        _searchBar.delegate = self;
        _searchBar.showsCancelButton = NO;
        _searchBar.backgroundImage = [UIImage imageWithColor:[UIColor whiteColor] size:_searchBar.bounds.size];
        [_searchBar setSearchFieldBackgroundPositionAdjustment:UIOffsetMake(0, 0)];
                
        UITextField *searchField = [_searchBar valueForKey:@"searchField"];
          if (searchField) {
                  if (@available(iOS 13.0, *)){
                      _searchBar.searchTextField.backgroundColor= COLOR_HEX(0xF2F2F2);
                  }else{
                      searchField.backgroundColor = COLOR_HEX(0xF2F2F2);
                  }

              searchField.layer.cornerRadius = kSearchBarHeight * 0.5;
              searchField.layer.masksToBounds = YES;
          }
    }
    return _searchBar;
}


- (NSMutableArray *)dataArray
{
    if (_dataArray == nil) {
        _dataArray = NSMutableArray.new;
    }
    return _dataArray;
}


- (NSMutableArray *)searchResults {
    if (_searchResults == nil) {
        _searchResults = NSMutableArray.new;
    }
    return _searchResults;
}

- (NSMutableArray *)members {
    if (_members == nil) {
        _members = NSMutableArray.new;
    }
    return _members;
}


@end
#undef kSearchBarHeight

File: AgoraChat-Swift/AgoraChat-Swift/code/Custom/ViewController/ACDSearchTableViewController.swift
//
//  ACDSearchTableViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/27.
//

import UIKit
import SnapKit

class ACDSearchTableViewController<T>: ACDBaseTableViewController, UISearchBarDelegate {

    let searchBar = UISearchBar()
    
    var isSearchState: Bool = false
    
    var searchResultHandle: (() -> Void)?
    var searchCancelHandle: (() -> Void)?
    
    var dataArray: [T] = []
    var searchSource: [IAgoraRealtimeSearch] = []
    var searchResults: [IAgoraRealtimeSearch] = []
    var page = 1
    var members: [String] = []
    var sectionTitles: [String] = []
    
    private let realtimeSearchUtil = AgoraRealtimeSearchUtils()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.searchBar.searchBarStyle = .minimal
        self.searchBar.backgroundColor = .white
        self.searchBar.placeholder = "Search".localized
        self.searchBar.delegate = self
        self.searchBar.showsCancelButton = false
        self.searchBar.searchFieldBackgroundPositionAdjustment = .zero
        if let textField = self.searchBar.value(forKey: "searchField") as? UITextField {
            if #available(iOS 13.0, *) {
                self.searchBar.searchTextField.backgroundColor = Color_F2F2F2
            } else {
                textField.backgroundColor = Color_F2F2F2
            }
            textField.layer.cornerRadius = 16
            textField.layer.masksToBounds = true
        }
        
        self.view.addSubview(self.searchBar)
        self.searchBar.snp.makeConstraints { make in
            make.top.left.right.equalTo(self.view)
        }
        self.table.snp.remakeConstraints { make in
            make.top.equalTo(self.searchBar.snp.bottom)
            make.left.right.equalTo(self.view)
            make.bottom.equalTo(self.view).offset(-5.0)
        }
    }
    
    func cancelSearchState() {
        self.searchBar.text = ""
        self.searchBar.setShowsCancelButton(false, animated: true)
        self.searchBar.resignFirstResponder()
        self.realtimeSearchUtil.cancel()
        self.isSearchState = false
        self.searchCancelHandle?()
        self.table.isScrollEnabled = !self.isSearchState
        self.table.reloadData()
    }
    
    func searchBarShouldBeginEditing(_ searchBar: UISearchBar) -> Bool {
        searchBar.setShowsCancelButton(true, animated: true)
        return true
    }

    func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
        self.isSearchState = true
        if self.searchBar.text == "" {
            self.table.isUserInteractionEnabled = false
            self.searchResults.removeAll()
            self.table.reloadData()
        } else {
            self.table.isUserInteractionEnabled = true
        }
    }
    
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
        searchBar.setShowsCancelButton(false, animated: true)
        self.table.isUserInteractionEnabled = true
        self.searchBar.resignFirstResponder()
    }

    func searchBar(_ searchBar: UISearchBar, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        return true
    }
    
    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        self.table.isUserInteractionEnabled = true
        self.searchResults.removeAll()
        if searchBar.text?.count ?? 0 <= 0 {
            self.table.reloadData()
            return
        }
        self.realtimeSearchUtil.realSearch(source: self.searchSource, keyword: searchText) { result in
            self.searchResults = result
            self.searchResultHandle?()
            self.table.reloadData()
        }
    }
    
    func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
    }

    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        self.cancelSearchState()
    }
}

extension ACDSearchTableViewController where T == [AgoraUserModel] {
    func sortAndReloadContacts(_ contacts: [String], subscribe: Bool = false) {
        if contacts.count <= 0 {
            self.dataArray.removeAll()
            self.sectionTitles.removeAll()
            self.searchSource.removeAll()
            self.table.reloadData()
            return
        }
        ContactUtil.sortContacts(contacts) { result, sectionTitles, searchSource in
            self.dataArray = result
            self.sectionTitles = sectionTitles
            self.searchSource = searchSource
            self.table.reloadData()
        }
        if subscribe {
            PresenceManager.shared.subscribe(members: contacts, completion: nil)
        }
    }
}
